source/extensions/filters/http/abac/abac_filter.cc

```diff
@@ -0,0 +1,171 @@
+public:
+  HeadersWrapper(const Http::HeaderMap& headers) : headers_(headers) {}
+
+  absl::optional<CelValue> operator[](CelValue key) const override {
+    if (!key.IsString()) {
+      return {};
+    }
+    auto out = headers_.get(Http::LowerCaseString(std::string(key.StringOrDie().value())));
+    if (out == nullptr) {
+      return {};
+    }
+    return CelValue::CreateString(out->value().getStringView());
+  }
+  int size() const override { return headers_.size(); }
+  bool empty() const override { return headers_.empty(); }
+  const CelList* ListKeys() const override { return nullptr; }
+  const Http::HeaderMap& headers_;
+};
+
+class RequestWrapper : public CelMap {
+public:
+  RequestWrapper(const Http::HeaderMap& headers, const StreamInfo::StreamInfo& info)
+      : wrapper_(headers), info_(info) {}
+
+  absl::optional<CelValue> operator[](CelValue key) const override {
+    if (!key.IsString()) {
+      return {};
+    }
+    auto value = key.StringOrDie().value();
+    if (value == kPath) {
+      return CelValue::CreateString(wrapper_.headers_.Path()->value().getStringView());
+    } else if (value == kHost) {
+      return CelValue::CreateString(wrapper_.headers_.Host()->value().getStringView());
+    } else if (value == kScheme) {
+      return CelValue::CreateString(wrapper_.headers_.Scheme()->value().getStringView());
+    } else if (value == kMethod) {
+      return CelValue::CreateString(wrapper_.headers_.Method()->value().getStringView());
+    } else if (value == kReferer) {
+      return CelValue::CreateString(wrapper_.headers_.Referer()->value().getStringView());
+    } else if (value == kHeaders) {
+      return CelValue::CreateMap(&wrapper_);
+    } else if (value == kTime) {
+      return CelValue::CreateTimestamp(absl::FromChrono(info_.startTime()));
+    }
+    return {};
+  }
+
+  int size() const override { return 0; }
+  bool empty() const override { return false; }
+  const CelList* ListKeys() const override { return nullptr; }
+
+private:
+  const HeadersWrapper wrapper_;
+  const StreamInfo::StreamInfo& info_;
+};
+
+} // namespace
+
+Http::FilterHeadersStatus AttributeBasedAccessControlFilter::decodeHeaders(Http::HeaderMap& headers,
+                                                                           bool) {
+  Protobuf::Arena arena;
+  google::api::expr::runtime::Activation activation;
+  activation.InsertValue(
+      kMetadata, CelValue::CreateMessage(&callbacks_->streamInfo().dynamicMetadata(), &arena));
+  const ConnectionWrapper connection(callbacks_->streamInfo());
+  activation.InsertValue(kConnection, CelValue::CreateMap(&connection));
+  const RequestWrapper request(headers, callbacks_->streamInfo());
+  activation.InsertValue(kRequest, CelValue::CreateMap(&request));
+
+  auto eval_status = expr_->Evaluate(activation, &arena);
+  if (!eval_status.ok()) {
+    ENVOY_LOG(debug, "evaluation failed: {}", eval_status.message());
+    return Http::FilterHeadersStatus::StopIteration;
```

Comment:
`StopAllIterationAndWatermark`

Reply:
Curious, why doesn't RBAC filter do that?

Reply:
ah nvm, this is local evaluation so it is fine.